keps/sig-api-machinery/4006-transition-spdy-to-websockets/README.md

```diff
@@ -0,0 +1,1200 @@
+
+As previously mentioned, we propose incrementing the subprotocol version from four
+to five: `v5.channel.k8s.io`. This version will indicate to upstream components
+that `kubectl` is interested in using WebSockets between the client and the API Server
+if it is supported.
+
+### Proposal: API Server `StreamTranslatorProxy`
+
+![Stream Translator Proxy](./stream-translator-proxy-2.png)
+
+Currently, the API Server role within client/container streaming is to proxy the
+data stream using the `UpgradeAwareProxy`. This enhancement proposes to modify the
+SPDY data stream between `kubectl` and the API Server by conditionally adding a
+`StreamTranslatorProxy` at the API Server. If the request is for a WebSocket upgrade
+of version `v5.channel.k8s.io`, the `UpgradeAwareProxy` will delegate to the
+`StreamTranslatorProxy`. This translation proxy terminates the WebSocket connection,
+and de-multiplexes the various streams in order to pass the data on to a SPDY connection,
+which continues upstream (to Kubelet and eventually the container runtime).
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+The goal is to ensure that we don't accept enhancements with inadequate testing.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+<!--
+Based on reviewers feedback describe what additional tests need to be added prior
+implementing this enhancement to ensure the enhancements have also solid foundations.
+-->
+
+```
+<<[UNRESOLVED]>>
+TBD - currently unimplemented
+<<[/UNRESOLVED]>>
+```
+
+##### Unit tests
+
+<!--
+In principle every added code should have complete unit test coverage, so providing
+the exact set of tests will not bring additional value.
+However, if complete unit test coverage is not possible, explain the reason of it
+together with explanation why this is acceptable.
+-->
+
+<!--
+Additionally, for Alpha try to enumerate the core package you will be touching
+to implement this enhancement and provide the current unit coverage for those
+in the form of:
+- <package>: <date> - <current test coverage>
+The data can be easily read from:
+https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+
+```
+<<[UNRESOLVED]>>
+TBD - currently unimplemented
+<<[/UNRESOLVED]>>
+```
+
+- `<package>`: `<date>` - `<test coverage>`
+
+##### Integration tests
```

Comment:
These are going to be important since those can allow us to exercise the fallback mechanism

Reply:
Agreed. I intend to submit separate PRs to outline the testing plan as well as the Production Readiness questions within the KEP. Submitting this initial iteration of the provisional KEP will allow others to more easily contribute to unimplemented sections.